ioctl: nvme_lm_cdq use nvme_passthru_cmd directly

ikegami-t opened the pull request:
Drop struct nvme_lm_cdq_args.

igaw committed:
ioctl: nvme_get_log use nvme_passthru_cmd directly

Refactor nvme_get_log and nvme_get_log_page to use use struct
nvme_passthru_cmd directly and drop the struct nvme_get_log_args.

The struct args have are ABI thus it's hard to change it after a
release. The nvme_passthru_cmd buffer is a generic buffer and it's
possible to use the setters directly on the buffer in an inline
function. Besides a reduced API to maintain it also generates better
code and avoids one call less when issue a command.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

merge with: 532c20955cd5 ("ioctl: nvme_get_log use nvme_passthru_cmd directly")

igaw committed:
ioctl: reorder nvme_get_log and nvme_get_log_page

The next change will add a static inline implementation of nvme_get_log
which uses nvme_get_log_page. Thus move the nvme_get_log_page function
up so it can be used directly.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

MaisenbacherD committed:
ioctl: nvme_set_features use nvme_passthru_cmd directly

Drop struct nvme_set_features_args.

Signed-off-by: Dennis Maisenbacher <dennis.maisenbacher@wdc.com>

igaw committed:
ioctl: nvme_identify use nvme_passthru_cmd directly

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: nvme_format_nvm use nvme_passthru_cmd directly

Drop struct nvme_format_nvm_args.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: nvme_ns_mgmt use nvme_passthru_cmd directly

Drop struct nvme_ns_mgmt_args.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: nvme_ns_attach use nvme_passthru_cmd directly

Drop struct nvme_ns_attach_args.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: nvme_ns_download use nvme_passthru_cmd directly

Drop struct nvme_fw_download_args.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: nvme_fw_commit use nvme_passthru_cmd directly

Drop struct nvme_fw_commit_args.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

igaw committed:
ioctl: nvme_security_send use nvme_passthru_cmd directly

Drop struct nvme_security_send_args.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_dsm use nvme_passthru_cmd directly

Drop struct nvme_dsm_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_io use nvme_passthru_cmd directly

Drop struct nvme_io_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ioctl: reorder nsid argument

Also order the pointer and its length member go to the end of the list.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: add dsm and io commands dword fields

To use the fields in nvme_io() and nvme_dsm() functions.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_copy use nvme_passthru_cmd directly

Drop struct nvme_copy_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

igaw committed:
test: fix rae and lsp handling in log page tests

The compiler warns with

../test/ioctl/logs.c: In function ‘test_get_log_reachability_groups’:
../test/ioctl/logs.c:725:28: warning: enum constant in boolean context [-Wint-in-bool-context]
  725 |                          (!!TEST_LSP << 8) | (!!TEST_RAE << 15) |
      |                            ^
../test/ioctl/logs.c:733:70: warning: enum constant in boolean context [-Wint-in-bool-context]
  733 |         err = nvme_get_log_reachability_groups(test_link, TEST_RAE, !!TEST_LSP,
      |                                                                      ^

Update the tests so the types are correctly handled. It also addresses
the problem with the wrong lsp value testing.

Signed-off-by: Daniel Wagner <wagi@kernel.org>

ikegami-t committed:
ioctl: nvme_resv_acquire use nvme_passthru_cmd directly

Drop struct nvme_resv_acquire_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_resv_register use nvme_passthru_cmd directly

Drop struct nvme_resv_register_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_resv_release use nvme_passthru_cmd directly

Drop struct nvme_resv_release_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_resv_report use nvme_passthru_cmd directly

Drop struct nvme_resv_report_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_io_mgmt_recv use nvme_passthru_cmd directly

Drop struct nvme_io_mgmt_recv_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_io_mgmt_send use nvme_passthru_cmd directly

Drop struct nvme_io_mgmt_send_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_zns_mgmt_send use nvme_passthru_cmd directly

Drop struct nvme_zns_mgmt_send_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_zns_mgmt_recv use nvme_passthru_cmd directly

Drop struct nvme_zns_mgmt_recv_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_zns_append use nvme_passthru_cmd directly

Drop struct nvme_zns_append_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_dim_send use nvme_passthru_cmd directly

Drop struct nvme_dim_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
fabrics: fix nvme_get_adrfam function description parameter ctrl as c

Since the warning message output by kernel-doc-check.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_lm_cdq use nvme_passthru_cmd directly

Drop struct nvme_lm_cdq_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t commented:
The PR changes for the work mentioned by the comment: https://github.com/linux-nvme/libnvme/issues/495#issuecomment-3173415434.
  Note: nvme_dim_send() already merged by the PR: https://github.com/nvme-experiments/libnvme/pull/4.

ikegami-t committed:
ioctl: nvme_lm_track_send use nvme_passthru_cmd directly

Drop struct nvme_lm_track_send_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_lm_migration_send use nvme_passthru_cmd directly

Drop struct nvme_lm_migration_send_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
ioctl: nvme_lm_migration_recv use nvme_passthru_cmd directly

Drop struct nvme_lm_migration_recv_args.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

igaw commented on src/nvme/ioctl.h:
```diff
@@ -5056,12 +5056,68 @@ static inline int nvme_lm_track_send(nvme_link_t l, __u16 mos, __u16 cdqid, __u8
  * nvme_lm_migration_send() - Migration Send command
  * @l:		Link handle
- * @args:	&struct nvme_lm_migration_send_args argument structure
+ * @offset:	Offset: This field specifies the offset, in bytes, within the data available to be
```
Let's rename it to `lpo`, so it matches with the other commands. And it should be before `data`.

ikegami-t replied:
Actually described as `Controller State Offset Lower/Upper (CSOL/U)` so better to rename as `cso`?
![migration-send-cso](https://github.com/user-attachments/assets/1efe0325-eba2-4936-851f-6f17c37e624f)

igaw replied:
yes, let's follow the spec here closely. That avoids misunderstanding (I thought `offset` was `lpo` but it isn't).

The arg should be be of type `u64` and should be called `cso`.

ikegami-t replied:
Noted.

igaw commented on src/nvme/ioctl.h:
```diff
@@ -5056,12 +5056,68 @@ static inline int nvme_lm_track_send(nvme_link_t l, __u16 mos, __u16 cdqid, __u8
+ * @dudmq:	Delete User Data Migration Queue (DUDMQ): If set, the migration queue is deleted
+ *		is deleted as part of the Suspend operation. If cleared, it is retained.
+ * @data:	Pointer to data
+ * @numd:	Number of Dwords (NUMD): This field specifies the number of dwords being transferred
```
need to check the spec here, but  I'd prefer to call it `len` and do the 0 index math inside the function. But it might be a special case here for the `nvme_lm_...` APIs.

ikegami-t replied:
The specification describes as `Number of Dwords (NUMD)` actually but still needed to call as `len`?
![migration-send-numd](https://github.com/user-attachments/assets/274fea3d-2fca-44e2-ab12-f181c4b8cc57)

igaw replied:
Do we actually need this parameter at all? I mean looking at the code the `numd` argument is used also for `data_len` (and actually incorrectly  (off by one) if I am not mistaken)

```c
	struct nvme_passthru_cmd cmd = {
		.opcode = nvme_admin_migration_send,
		.cdw10 = cdw10,
		.cdw11 = cdw11,
		.cdw12 = (__u32)args->offset,
		.cdw13 = (__u32)(args->offset >> 32),
		.cdw14 = NVME_SET(args->uidx, LM_MIGRATION_SEND_UIDX),
		.cdw15 = args->numd,
		.addr = (__u64)(uintptr_t)args->data,
		.data_len = args->numd << 2,
		.timeout_ms = args->timeout,
	};
```

If we have a `data` pointer we also should have a `len` argument to match the existing interfaces. I think when they defined the migration interface they moved the existing `numd` to a new `cdw` so it fits into one word instead splitting into two `cdw`

ikegami-t replied:
Understood so I will change the paramter to len (and set the NUMD field by using the len parameter value). The NUMD field is not zero based value as not described (and described as 0h value set for the sequence indicator field: 10b as the last command) so looks correctly used I think.
![migration-send-seqind](https://github.com/user-attachments/assets/8b8d3603-147d-4c7f-98e4-650d6cc971e5)

igaw commented on src/nvme/ioctl.h:
```diff
@@ -5122,12 +5122,56 @@ static inline int nvme_lm_migration_send(nvme_link_t l, __u64 offset, __u16 mos,
- * @args:	&struct nvme_lm_migration_rev_args argument structure
+ * @offset:	Offset: This field specifies the offset, in bytes, within the data available to be
+ *		returned and specifies the starting point for that data for what is actually
+ *		returned to the host.
```
same here this is `lpo`.

ikegami-t replied:
It is described actually as `Offset Lower/Upper (OL/U)` so seems the `offset` is correct naming.
![migration-receive-o](https://github.com/user-attachments/assets/408cbbdb-ac21-4b08-89b0-aac19dd10f0d)

igaw replied:
ah okay, so `offset` would be correct here.

ikegami-t replied:
I see.

igaw commented on src/nvme/ioctl.h:
```diff
@@ -5122,12 +5122,56 @@ static inline int nvme_lm_migration_send(nvme_link_t l, __u64 offset, __u16 mos,
+ * @csuidxp:	Controller State UUID Index Parameter (CSUIDXP): This field is vendor specific.
+ * @data:	Pointer to data
+ * @numd:	Number of Dwords (NUMD): This field specifies the number of dwords to return. This
+ *		is a 0's based value.
```
again need to check the spec what's this numd thing means in this context. It looks like the `len` argument from the get log page function but who knows :)

ikegami-t replied:
Described as `Number of Dwords (NUMDL)`.
![migration-receive-numdl](https://github.com/user-attachments/assets/36c25209-1653-4620-acce-511ad9e218fd)

igaw replied:
I am pretty sure this is the `len` argument. I'll ask someone who should know this detail.

ikegami-t replied:
Yes I will do change to use len argument and set the NUMDL field by using the argument.

igaw commented:
I'm on vacation till Tuesday, so I'd like not do any review etc in the next view days.

ikegami-t commented:
I also will go travel tomorrow until Sunday so will continue to work after that. Thank you.

ikegami-t committed:
test/zns: call nvme_zns_report_zones() without timeout parameter

Since the function parameter deleted.

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t committed:
test/zns: fix kernel-doc-check warnings

Since the following warnings output.

$ ./scripts/kernel-doc-check test/zns.c
test/zns.c:3: warning: This comment starts with '/**', but isn't a kernel-doc comment. Refer Documentation/doc-guide/kernel-doc.rst
test/zns.c:10: warning: This comment starts with '/**', but isn't a kernel-doc comment. Refer Documentation/doc-guide/kernel-doc.rst

Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

ikegami-t commented:
Fixed the changes as mentioned.

igaw commented:
I've merged them now (reordered a couple of patches, merged the fixup patch) I haven't reviewed them in details. There are a few minor things which I'll address later. Let's get first the ioctl change done and then update nvme-cli accordingly, so we can start testing it with blktests.

igaw commented:
Thanks!
